Implement merging fonts to beyond 64k glyphs

behdad opened the pull request.

behdad committed:
[>64k:glyf+maxp] Support >64k gids in composite glyphs and in maxp table

This implements part of https://github.com/be-fonts/boring-expansion-spec/issues/42

behdad committed:
[>64k:hmtx] Implement >64k glyphs

Implements https://github.com/be-fonts/boring-expansion-spec/issues/7

Optimization not implemented.

behdad committed:
[>64k:merger:cmap] Support >64k

By not generating a format4 subtable.

behdad committed:
[>64k:merger:post] Support >64k

By dropping glyph-name encoding.

behdad committed:
[>64k:ClassDef/Coverage] Support >64k

Implements https://github.com/be-fonts/boring-expansion-spec/issues/30

behdad committed:
[>64k:merger:cmap+post] Adjust condition to use glyphOrder

behdad committed:
[>64k:layout:PairPos] Implement PairPos format 3/4

And merger upgrade logic.

Implements https://github.com/be-fonts/boring-expansion-spec/issues/38

behdad committed:
[merge] Separate classify_context

behdad committed:
[>64k:layout:(Chain)Context] Implement, and add merger upgrade logic

Implements https://github.com/be-fonts/boring-expansion-spec/issues/34

behdad committed:
[>64k:layout:Ligature] Implement and add merger upgrade logic

Implements https://github.com/be-fonts/boring-expansion-spec/issues/33

behdad committed:
[>64k:layout:SingleSubst] Implement and add merger upgrade logic

Implements https://github.com/be-fonts/boring-expansion-spec/issues/31

behdad committed:
[>64k:layout:MultipleSubst] Implement and add merger upgrade logic

Implements https://github.com/be-fonts/boring-expansion-spec/issues/32

behdad committed:
[>64k:layout:AlternateSubst] Implement & add merger upgrade logic

Implements https://github.com/be-fonts/boring-expansion-spec/issues/32

behdad committed:
[otBase] Fix sortCoverageLast logic

I'm not sure why no one hit the problem with it before me now.
Or rather, I'm not sure why I'm hitting it now.  But the fix is
legit.

behdad committed:
[>64k:layout:MarkBase/MarkMark/MarkLigPos] Implement & merger upgrade

Implements https://github.com/be-fonts/boring-expansion-spec/issues/40

behdad committed:
[>64k:layout] GSUB/GPOS is complete now

behdad committed:
Make tests pass

behdad committed:
[>64k:merger] Adjust logic

No need to check. If we are in this function, upgrade is advised.

behdad committed:
[otTables] Upgrade all lookups to Extension if more than 128

This takes care of the long loop of Extension upgrades in
huge fonts.

behdad committed:
[otConverters] Add Offset32

behdad commented:
I'm now hitting `LookupList` offset overflows now, which means I need to bump GSUB/GPOS structs to version 2.0. That's a bit hard to do in ttx currently.

behdad committed:
[>64k:layout:(Chain)Context] Add format 5/6

To extend offset reach.

https://github.com/be-fonts/boring-expansion-spec/issues/34

behdad committed:
[>64k:layout] Add GSUB/GPOS version 2.0

To fix offset overflows of LookupList itself.

Implements https://github.com/be-fonts/boring-expansion-spec/issues/58

behdad committed:
[>64k:glyf] Support decoding 24bit GlyphID in composites

Part of https://github.com/be-fonts/boring-expansion-spec/issues/42

behdad committed:
[>64k:ttFont] Support loading numGlyphs from loca table

behdad committed:
[>64k:layout:ClassDef/Coverage] Implement reading and fix writing

Ouch, my writing logic for all tables is flawed. Fixing one by one.

behdad committed:
[>64k:layout:MultipleSubst/AlternateSubst] Implement read / fix write

behdad committed:
[>64k:layout:SingleSubst] Support reading

behdad committed:
[>64k:layout:LigatureSubst] Implement read & fix writing

behdad committed:
[>64k:layout:PairPos] Fix writing

behdad committed:
[>64k:layout:(Chain)Context] Fix writing

behdad committed:
Fix test

behdad committed:
[otTables] Downgrade an error message to info

behdad committed:
[>64k:layout] Downgrade Offset32's to Offset24

This is enough and makes implementation more uniform.

Already updated the spec pages.

behdad committed:
[>64k:layout] Downgrade most counts to 16bit again

Did this already in the respective spec issues.

The main issue was that the RangeRecord was pushed to 24bit values
unnecessarily in most cases.

anthrotype commented:
I'm away and won't be able to review this before I'll be back to work on 18 July, sorry!

behdad committed:
[>64k:layout] Upgrade offsets to 24bit in (Chain)Context

Updated spec in https://github.com/be-fonts/boring-expansion-spec/issues/34

behdad committed:
[>64k:layout] Remove (Chain)Context format 6

Not necessary. Spec updated.

https://github.com/be-fonts/boring-expansion-spec/issues/34

behdad committed:
[>64k:layout] Upgrade main GSUB/GPOS structure offsets to 24bit

Spec updated: https://github.com/be-fonts/boring-expansion-spec/issues/58

behdad committed:
[>64k:layout] Fix MultipleSubstFormat2 layout

https://github.com/be-fonts/boring-expansion-spec/issues/32

behdad committed:
[>64k:hmtx] Implement decompile

Implements https://github.com/be-fonts/boring-expansion-spec/issues/7

behdad committed:
[>64k] Handle lsb=None when drawing glyphs

khaledhosny commented:
> I don't know if I'm 100% comfortable with the latter idea

Why not?

khaledhosny commented:
> > I don't know if I'm 100% comfortable with the latter idea
> 
> Why not?

I mean, HarfBuzz supported is merged in main already and will be available in the next release.

behdad commented:
> before I begin to review, I was wondering what's the plan with this. Are you planning on keeping it as a separate branch (occasionally rebased on main) while the work progresses on the boring-expansion spec front, or do you intend to have it in main before any of it becomes official in the OT spec? I don't know if I'm 100% comfortable with the latter idea...

As Khaled said, this is already in HarfBuzz main branch. So in that sense it's ahead of where COLRv1 was when it got merged into fonttools. As for OT spec, it's likely that this work will never make it into OT spec, because it's quite likely that MS will not have resources to implement it. Implementing it in HarfBuzz was a week's work because of the HarfBuzz architecture that allowed templatizing types. MS implementation does not afford them the same from what I understand.

So, this is a HB-only font format for now. Next step is to generate a single-page spec.

behdad committed:
[>64k:layout] Upgrade main GDEF structure offsets to 24bit

Spec updated: https://github.com/be-fonts/boring-expansion-spec/issues/36

behdad committed:
[>64k:hhead] Handle lsb=None

anthrotype commented:
I'd suggest that you first `git rebase 78a939848` (78a939848 is the commit on main that immediately precedes the black PR #2925), fix the merge conflicts, and then rebase on, or merge in the `main` branch, resolving any conflicts by automatically running black

behdad commented:
Superseded by https://github.com/fonttools/fonttools/pull/4097
